File: codes/mlp/Fortran/ext_module.f90



! Neural networks pedagogical materials
! The following code is free to use and modify to any extent 
! (with no responsibility of the original author)

! Reference to the author is a courtesy
! Author : David Cornu => david.cornu@utinam.cnrs.fr




module ext_module
	
	implicit none
	
	contains	
	

	!######################## ########################## ##########################
	
	!    						 ON-LINE FUNCTIONS
	
	!######################## ########################## ##########################
	
		
		subroutine forward(input, in_dim, hidden, hid_dim, output, out_dim, weights1, weights2, beta)
			!######################## ##########################
			!    Forward the all netork with logistic neurons
			!######################## ##########################
			real, dimension(:), intent(IN) :: input
			real, dimension(:), intent(INOUT) :: hidden
			real, dimension(:), intent(INOUT) :: output
			real, intent(IN) :: beta
			integer, intent(IN) :: in_dim, hid_dim, out_dim
			real, dimension(:,:), intent(IN) :: weights1, weights2
			
			real :: h
			integer :: i,j
			
			do i=1, hid_dim
				h = 0.0
				do j=1, in_dim+1
					h = h + weights1(j,i)*input(j)
				end do
				hidden(i) = 1.0/(1.0 + exp(-beta*h))
			end do
			hidden(hid_dim+1) = -1.0

			do i=1, out_dim
				h = 0.0
				do j=1, hid_dim+1
					h = h + weights2(j,i)*hidden(j)
				end do
				output(i) = 1.0/(1.0 + exp(-beta*h))
			end do
		
		end subroutine !forward
		
			
		
		
		
		
		subroutine backprop(input, in_dim, hidden, hid_dim, output, targ, out_dim, weights1, weights2, learn_rate, beta)
			!######################## ##########################
			! Backward on the all network with logistic neurons
			!######################## ##########################
		
			real, dimension(:), intent(IN) :: input, hidden, output, targ
			integer, intent(IN) :: in_dim, hid_dim, out_dim
			real, dimension(:,:), intent(INOUT) :: weights1, weights2
			real, intent(IN) :: learn_rate, beta
			
			real, dimension(out_dim) :: delta_o
			real, dimension(hid_dim) :: delta_h
			integer :: i, j
			real :: h
		
			do i=1, out_dim
				delta_o(i) = beta*(output(i)-targ(i))*output(i)*(1.0-output(i))
			end do

			do i=1, hid_dim
				h = 0.0
				do j=1, out_dim
					h = h + weights2(i,j)*delta_o(j)
				end do
				delta_h(i) = beta*hidden(i)*(1.0-hidden(i))*h
			end do


			do i=1, in_dim+1
				do j=1, hid_dim
					weights1(i,j) = weights1(i,j) - learn_rate*(delta_h(j)*input(i))
				end do
			end do
			
			do i=1, hid_dim+1
				do j=1, out_dim
					weights2(i,j) = weights2(i,j) - learn_rate*(delta_o(j)*hidden(i))
				end do
			end do
		
		
		end subroutine !backward
		
		
	
	
	
		subroutine confmat(input, in_dim, hid_dim, targ, out_dim, nb_data, weights1, weights2, beta)
			!######################## ##########################
			! Forward on an epoch and display a confusion matrix
			!######################## ##########################
			
			real, dimension(:,:), intent(IN) :: input, targ
			integer, intent(IN) :: in_dim, hid_dim, out_dim, nb_data
			real, dimension(:,:), intent(IN) :: weights1, weights2
			real, intent(IN) :: beta
			
			real, dimension(out_dim) :: output
			real, dimension(hid_dim+1) :: hidden
			integer, dimension(1) :: max_a, max_b
			integer, dimension(out_dim,out_dim) :: confmatrix
			real, dimension(out_dim) :: recall, precis
			real :: accu, quad_error
			
			integer :: i, j
			
			accu = 0
			confmatrix = 0
			quad_error = 0
			do i=1, nb_data
			
				call forward(input(i,:), in_dim, hidden, hid_dim, output, out_dim, weights1, weights2, beta)
				
				quad_error = quad_error + 0.5*sum((output(:) - targ(i,:))**2)
				
				max_a = maxloc(output)
				max_b = maxloc(targ(i,:))
				confmatrix(max_b(1), max_a(1)) = confmatrix(max_b(1), max_a(1)) + 1
				if(max_a(1) == max_b(1)) then
					accu = accu + 1
				end if
			end do
			
			do i=1, out_dim
				recall(i) = 0
				precis(i) = 0
				do j=1, out_dim
					recall(i) = recall(i) + confmatrix(i,j)
					precis(i) = precis(i) + confmatrix(j,i)
				end do
				if(recall(i) > 0.0)then
					recall(i) = confmatrix(i,i) / recall(i) * 100.0
				end if
				if(precis(i) > 0.0)then
					precis(i) = confmatrix(i,i) /precis(i) * 100.0
				endif
			end do
				
			write(*,*) "*****************************************************************"
			write(*,'(a11,a44)') "Confmat :","Recall"
			do i=1, out_dim
				write(*,'(a10)', advance="no") "         "
				do j=1, out_dim
					write(*,'(I10)', advance="no") confmatrix(i,j)
				end do
				write(*,'(a12,f6.2)') "     ", recall(i)
			end do		
			write(*,*)
			write(*,'(a13)',advance="no") "Precision"
			do i=1, out_dim
				write(*,'(f10.2)', advance="no") precis(i)
			end do
			write(*,'(a7,f8.2)', advance="no") "Accu" , real(accu)/real(nb_data)*100.0
			write(*,*)
			write(*,*) "*****************************************************************"
			write(*,*) "Average test set quadratic error : ", quad_error/real(nb_data)
			
		end subroutine !confmat
		
		
		
	
	
		
		subroutine shuffle(input, in_dim, targ, out_dim, nb_data)
			!######################## ##########################
			!               Fisher Yates Shuffle
			!######################## ##########################
			
			real, dimension(:,:), intent(INOUT) :: input, targ
			integer, intent(IN) :: in_dim, out_dim, nb_data
			
			real :: rdm
			integer :: i, ind
			real, allocatable :: temp_input(:), temp_targ(:)
			
			allocate(temp_input(in_dim))
			allocate(temp_targ(out_dim))
			
			do i=1, nb_data-1
				call random_number(rdm)
				ind = int(rdm * (nb_data-i)) + i
				
				temp_input(:) = input(i,:)
				input(i,:) = input(ind,:)
				input(ind,:) = temp_input(:)

				temp_targ(:) = targ(i,:)
				targ(i,:) = targ(ind,:)
				targ(ind,:) = temp_targ(:)

			end do
			
			deallocate(temp_input)
			deallocate(temp_targ)
		
		end subroutine !shuffle
		
		
		
		
	!######################## ########################## ##########################
	
	!    						   BATCH FUNCTIONS
	
	!######################## ########################## ##########################
	
	


		subroutine forward_batch(input, in_dim, hidden, hid_dim, output, out_dim, weights1, weights2, beta)
			!######################## ##########################
			!    Forward the all netork with logistic neurons
			!######################## ##########################
			real, dimension(:,:), intent(IN) :: input
			real, dimension(:,:), intent(INOUT) :: hidden
			real, dimension(:,:), intent(INOUT) :: output
			real, intent(IN) :: beta
			integer, intent(IN) :: in_dim, hid_dim, out_dim
			real, dimension(:,:), intent(IN) :: weights1, weights2
			
			real :: h
			integer :: i,j
			
			hidden = matmul(input, weights1)
			
			hidden(:,1:hid_dim) = 1.0/(1.0 + exp(-beta*hidden(:,1:hid_dim)))
			
			output = matmul(hidden, weights2)

			output(:,:) = 1.0/(1.0 + exp(-beta*output(:,:)))
		
		end subroutine !forward
		


		subroutine backprop_batch(input, in_dim, hidden, delta_h, hid_dim, output, &
			delta_o, targ, out_dim, weights1, weights2, learn_rate, beta)
			!######################## ##########################
			! Backward on the all network with logistic neurons
			!######################## ##########################
		
			real, dimension(:,:), intent(INOUT) :: input, hidden, delta_h, output, delta_o, targ
			integer, intent(IN) :: in_dim, hid_dim, out_dim
			real, dimension(:,:), intent(INOUT) :: weights1, weights2
			real, intent(IN) :: learn_rate, beta
			
			integer :: i, j
			real :: h
		
			delta_o(:,:) = beta*(output(:,:)-targ(:,:))*output(:,:)*(1.0-output(:,:))


			delta_h = matmul(delta_o, transpose(weights2))

			delta_h(:,1:hid_dim) = beta*hidden(:,1:hid_dim)*(1.0-hidden(:,1:hid_dim))*delta_h(:,1:hid_dim) 
			delta_h(:,hid_dim+1) = 0.0


			weights2 = weights2 - learn_rate*matmul(transpose(hidden),delta_o)
			weights1 = weights1 - learn_rate*matmul(transpose(input),delta_h)
		
		end subroutine !backward




		subroutine confmat_batch(input, in_dim, hidden, hid_dim, output, targ, out_dim, nb_data, weights1, weights2, beta)
			!######################## ##########################
			! Forward on an epoch and display a confusion matrix
			!######################## ##########################
			
			real, dimension(:,:), intent(INOUT) :: input, output, targ, hidden
			integer, intent(IN) :: in_dim, hid_dim, out_dim, nb_data
			real, dimension(:,:), intent(IN) :: weights1, weights2
			real, intent(IN) :: beta
			
			real, dimension(out_dim) :: temp_targ
			integer, dimension(1) :: max_a, max_b
			integer, dimension(out_dim,out_dim) :: confmatrix
			real, dimension(out_dim) :: recall, precis
			real :: accu, quad_error
			
			integer :: i, j
			
			accu = 0
			confmatrix = 0
			quad_error = 0

			call forward_batch(input, in_dim, hidden, hid_dim, output, out_dim, weights1, weights2, beta)
				
			quad_error = quad_error + 0.5*sum((output(:,:) - targ(:,:))**2)
				
			do i = 1, nb_data
				max_a = maxloc(output(i,:))
				max_b = maxloc(targ(i,:))
				confmatrix(max_b(1), max_a(1)) = confmatrix(max_b(1), max_a(1)) + 1
				if(max_a(1) == max_b(1)) then
					accu = accu + 1
				end if
			end do
			
			do i=1, out_dim
				recall(i) = 0
				precis(i) = 0
				do j=1, out_dim
					recall(i) = recall(i) + confmatrix(i,j)
					precis(i) = precis(i) + confmatrix(j,i)
				end do
				if(recall(i) > 0.0)then
					recall(i) = confmatrix(i,i) / recall(i) * 100.0
				end if
				if(precis(i) > 0.0)then
					precis(i) = confmatrix(i,i) /precis(i) * 100.0
				endif
			end do
				
			write(*,*) "*****************************************************************"
			write(*,'(a11,a44)') "Confmat :","Recall"
			do i=1, out_dim
				write(*,'(a10)', advance="no") "         "
				do j=1, out_dim
					write(*,'(I10)', advance="no") confmatrix(i,j)
				end do
				write(*,'(a12,f6.2)') "     ", recall(i)
			end do		
			write(*,*)
			write(*,'(a13)',advance="no") "Precision"
			do i=1, out_dim
				write(*,'(f10.2)', advance="no") precis(i)
			end do
			write(*,'(a7,f8.2)', advance="no") "Accu" , real(accu)/real(nb_data)*100.0
			write(*,*)
			write(*,*) "*****************************************************************"
			write(*,*) "Average test set quadratic error: ", quad_error/real(nb_data)
			
		end subroutine !confmat






end module

File: codes/mlp/C/ext_module.c


/* 
Neural networks pedagogical materials
The following code is free to use and modify to any extent (with no responsibility of the original author)

Reference to the author is a courtesy
Author : David Cornu => david.cornu@utinam.cnrs.fr
*/


#include <stdio.h>
#include <stdlib.h>
#include <tgmath.h>

float **create_2d_table(int m, int n)
{
	int i;
	float** tab;
	float* temp_tab;
	
	tab = (float**) malloc(m*sizeof(float*));
	temp_tab = (float*) malloc(m*n*sizeof(float));
	
	for(i = 0; i < m; i++)
		tab[i] = &temp_tab[i*n];
		
	return tab;
}



int argmax(float *tab, int size)
{
	int i;
	float max;
	int imax;

	max = *tab;
	imax = 0;
	
	
	for(i = 1; i < size; i++)
	{
		if(tab[i] >= max)
		{
			max = tab[i];
			imax = i;
		}
	}
	
	return imax;
}



void forward(float *input, int in_dim, float* hidden, int hid_dim, float *output, int out_dim, float **weights1, float **weights2, float beta)
{
//######################## ##########################
//        One forward step with a logistic neuron
//######################## ##########################

	float h;
	int i,j;

	for(i = 0; i < hid_dim; i++)
	{
		h = 0.0;
		for(j = 0; j < in_dim + 1; j++)
			h += weights1[j][i]*input[j];
		hidden[i] = 1.0 / (1.0 + expf(-beta*h));
	}
	hidden[hid_dim] = -1.0;

	for(i = 0; i < out_dim; i++)
	{
		h = 0.0;
		for(j = 0; j < hid_dim + 1; j++)
			h += weights2[j][i]*hidden[j];
		output[i] = 1.0 / (1.0 + expf(-beta*h));
	}
	

}


void backprop(float *input, int in_dim, float *hidden, int hid_dim, float *output, float *targ, int out_dim, float **weights1, float **weights2, float learn_rate, float beta)
{
//######################## ##########################
//       One backward step with a logistic neuron
//######################## ##########################
	float delta_o[out_dim], delta_h[hid_dim+1];
	
	int i, j;
	float h;
	
	for(i = 0; i < out_dim; i++)
		delta_o[i] = beta*(output[i] - targ[i])* output[i] * (1.0 - output[i]);
	
	for(i = 0; i < hid_dim; i++)
	{
		h = 0.0;
		for(j = 0; j < out_dim; j++)
			h += weights2[i][j]*delta_o[j];
		delta_h[i] = beta*hidden[i]*(1.0 - hidden[i])*h;
	}
	
	for(i = 0; i < in_dim + 1; i++)
		for(j = 0 ; j < hid_dim; j++)
			weights1[i][j] -= learn_rate*delta_h[j]*input[i];
			
	for(i = 0; i < hid_dim + 1; i++)
		for(j = 0 ; j < out_dim; j++)
			weights2[i][j] -= learn_rate*delta_o[j]*hidden[i];
}


void confmat(float **input, int in_dim, int hid_dim, float **targ, int out_dim, int nb_data, float **weights1, float **weights2, float beta)
{
//######################## ##########################
// Forward on an epoch and display a confusion matrix
//######################## ##########################
	
	float output[out_dim];
	int max_a, max_b;
	int confmatrix[out_dim][out_dim];
	float recall[out_dim], precis[out_dim];
	float accu, quad_error;
	float hidden[hid_dim+1];
	
	int i, j;
	
	accu = 0.0;
	for(i = 0; i < out_dim*out_dim; i++)
		confmatrix[0][i] = 0.0;
		
	quad_error = 0.0;
	for(i = 0; i < nb_data; i++)
	{
		forward(input[i], in_dim, hidden, hid_dim, output, out_dim, weights1, weights2, beta);
		
		for(j = 0; j < out_dim; j++) 
			quad_error += (output[j] - targ[i][j])*(output[j] - targ[i][j]); 
		max_a = argmax(output, out_dim);
		max_b = argmax(targ[i], out_dim);
		confmatrix[max_b][max_a] += 1;
		if(max_a == max_b)
			accu = accu + 1;
	}
	
	for(i = 0; i < out_dim; i++)
	{
		recall[i] = 0;
		precis[i] = 0;
		for(j = 0; j < out_dim; j++)
		{
			recall[i] += confmatrix[i][j];
			precis[i] += confmatrix[j][i];
		}
		if(recall[i] > 0.0)
			recall[i] = confmatrix[i][i] / recall[i] * 100.0;

		if(precis[i] > 0.0)
			precis[i] = confmatrix[i][i] /precis[i] * 100.0;

	}
		
	printf("*****************************************************************\n");
	printf("Confmat :                                           Recall\n");
	for(i = 0; i < out_dim; i++)
	{
		printf("         ");
		for(j = 0; j < out_dim; j++)
			printf("%10d", confmatrix[i][j]);
		
		printf("        %6.2f\n", recall[i]);
	}	
	printf("\n  Precision");
	for(i=0; i < out_dim; i++)
		printf("%10.2f", precis[i]);

	printf(" Accu  %6.2f" , ((float)accu/(float)nb_data)*100.0);
	printf("\n*****************************************************************\n");
	printf("Average test set quadratic error: %f\n", quad_error/(float)nb_data);
}



void shuffle(float** input, int in_dim, float** targ, int out_dim, int nb_data)
{
//######################## ##########################
//               Fisher Yates Shuffle
//######################## ##########################
	
	int i, j, ind;
	float temp;
	
	for(i=0; i < nb_data-1 ; i++)
	{	
		
		ind = (int)(((double)rand()/(double)RAND_MAX) * (nb_data-i)) + i;
		
		for(j = 0; j < in_dim; j++)
		{
			temp = input[i][j];
			input[i][j] = input[ind][j];
			input[ind][j] = temp;
		}
		
		for(j = 0; j < out_dim; j++)
		{
			temp = targ[i][j];
			targ[i][j] = targ[ind][j];
			targ[ind][j] = temp;
		}
	}

}





































